src/connection_pool.rs

```diff
@@ -1077,3 +1138,118 @@ where
         connections.swap(i, new_idx);
     }
 }
+
+/// Start a thread to run periodic health checks on the connection pool
+fn start_connection_check<C>(
+    timer: &timer::Timer,
+    conn_check_interval: u64,
+    protected_data: ProtectedData<C>,
+    rebalance_check: RebalanceCheck,
+    log: Logger,
+) -> Guard
+where
+    C: Connection,
+{
+    debug!(
+        log,
+        "starting connection health task, interval {} seconds",
+        conn_check_interval
+    );
+    timer.schedule_repeating(
+        Duration::seconds(conn_check_interval as i64),
+        move || {
+            check_pool_connections(
+                protected_data.clone(),
+                rebalance_check.clone(),
+                log.clone(),
+            )
+        },
+    )
+}
+
+fn check_pool_connections<C>(
+    protected_data: ProtectedData<C>,
+    rebalance_check: RebalanceCheck,
+    log: Logger,
+) where
+    C: Connection,
+{
+    let mut connection_data = protected_data.connection_data_lock();
+    let len = connection_data.connections.len();
+
+    let mut rebalance = rebalance_check.get_lock();
+    if *rebalance {
+        debug!(
+            log,
+            "rebalance in progress, not proceeding with connection checks"
+        );
+        return;
+    }
+
+    if len == 0 {
+        debug!(log, "No connections to check.");
+        return;
```

Comment:
Should this signal the rebalancer to attempt to do work? Thinking specifically about the case where the pool is created, the service is at least partially up where to where the Resolver sees it and notifies the pool. For some reason the pool cannot immediately establish connections to the service, but at some later time the service is ready to accept connections. I don't know how else besides this check we would eventually establish those connections.